### PR TITLE
Feature/cu hk29r7 Distance model OpenAl and Wav to Ogg encoder

### PR DESCRIPTION
**Pull request by Floren94**
Distance Attenuation is working for Wav Mono files when Spatial Blend is set to 3D.
Captura1
Captura2

Audio Importer can convert Wav to Ogg.

### CODE REVIEW

**Review by FrancPS — COMMENTED** [review]
As usual, check the feature in a test scene in GAME configuration.

Also, any new class members that are "configurations" should be serialized.

File: Project/Source/Components/ComponentAudioListener.cpp [hunk 1]
```diff
@@ -72,6 +79,26 @@ void ComponentAudioListener::UpdateAudioListener() {
...
+		if (clamped)
+			alDistanceModel(AL_INVERSE_DISTANCE_CLAMPED);
+		else
+			alDistanceModel(AL_INVERSE_DISTANCE);
+		break;
+	case DistanceModel::LINEAR:
+		if (clamped)
+			alDistanceModel(AL_LINEAR_DISTANCE_CLAMPED);
+		else
+			alDistanceModel(AL_EXPONENT_DISTANCE);
```
Comment (JaimeBea) [inline Project/Source/Components/ComponentAudioListener.cpp:99]: If/else statements should always use braces

File: Project/Source/FileSystem/AudioImporter.h [hunk 2]
```diff
@@ -1,7 +1,16 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+#include <sndfile.h>
+
+#define BUFFER_LEN 4096
 
 namespace AudioImporter {
 	bool ImportAudio(const char* filePath, JsonValue jMeta);
+
+	static void EncondeWavToOgg(const char* infilename, const char* outfilename);
```
Comment (JaimeBea) [inline Project/Source/FileSystem/AudioImporter.h:15]: Static within a namespace is redundant, you can remove the static keyword.

File: Project/Source/FileSystem/AudioImporter.cpp [hunk 3]
```diff
@@ -51,4 +66,41 @@ bool AudioImporter::ImportAudio(const char* filePath, JsonValue jMeta) {
...
 	LOG("Audio imported in %ums", timeMs);
 	return true;
+}
+
+void AudioImporter::EncondeWavToOgg(const char* infilename, const char* outfilename) {
+	static double data[BUFFER_LEN];
+
+	SNDFILE *infile, *outfile;
+	SF_INFO sfinfo;
+	int readcount;
```
Comment (JaimeBea) [inline Project/Source/FileSystem/AudioImporter.cpp:76]: Code Style:
- Variables should be named using camelCase:
  - inFile, outFile, inFileName...

**Review by JaimeBea — APPROVED** [review]

File: Project/Source/Components/ComponentAudioSource.h [hunk 4]
```diff
@@ -42,4 +42,7 @@ class ComponentAudioSource : public Component {
 	float innerAngle = 90.f;  // Inner angle from directional source where gain parameter will be "gain"
 	float outerAngle = 180.f; // Outer angle from directional source where gain parameter will be "outerGain"
 	float outerGain = 0.f;	  // Gain applied outside outerAngle. Between inner and outer, OpenAL do an interpolation
+	float rollOffFact = 1.0f;
+	float referenceDistance = 6.0f;
+	float maxDistance = 15.0f;
```
Comment (JaimeBea) [inline Project/Source/Components/ComponentAudioSource.h:47]: Missing serialization

**Review by JaimeBea — CHANGES_REQUESTED** [review]

**Review by JaimeBea — DISMISSED** [review]

File: Project/Source/Components/ComponentAudioListener.cpp (hunk 1 above)
Comment (Floren94) [reply Project/Source/Components/ComponentAudioListener.cpp:101]: braces included

File: Project/Source/Components/ComponentAudioSource.h (hunk 4 above)
Comment (Floren94) [reply Project/Source/Components/ComponentAudioSource.h:47]: Serialization done

**Review by JaimeBea — DISMISSED** [review]

File: Project/Source/Components/ComponentAudioSource.cpp [hunk 5]
```diff
@@ -156,6 +159,12 @@ void ComponentAudioSource::OnEditorUpdate() {
 
 	ImGui::Separator();
 
+	ImGui::DragFloat("Roll Off Factor", &rollOffFact, 1.f, 0.0f, inf);
```
Comment (lowysole) [inline Project/Source/Components/ComponentAudioSource.cpp:162]: Same from ComponentListener. In this case keep the functions on UpdateSourceParameters()

File: Project/Source/FileSystem/AudioImporter.cpp [hunk 6]
```diff
@@ -51,4 +66,41 @@ bool AudioImporter::ImportAudio(const char* filePath, JsonValue jMeta) {
...
+		return;
+	};
+
+	if (!(outFile = sf_open(outfilename, SFM_WRITE, &sfInfo))) {
+		printf("Error : could not open file : %s\n", outfilename);
+		puts(sf_strerror(NULL));
+		exit(1);
+	};
+	
+	while ((readcount = sf_read_double(inFile, data, BUFFER_LEN))) {
```
Comment (lowysole) [inline Project/Source/FileSystem/AudioImporter.cpp:98]: readCount

File: Project/Source/FileSystem/AudioImporter.cpp [hunk 7]
```diff
@@ -51,4 +66,41 @@ bool AudioImporter::ImportAudio(const char* filePath, JsonValue jMeta) {
...
+		sf_write_double(outFile, data, readcount);
```
Comment (lowysole) [inline Project/Source/FileSystem/AudioImporter.cpp:99]: readCount

File: Project/Source/FileSystem/AudioImporter.cpp [hunk 8]
```diff
@@ -13,13 +13,28 @@
...
 	// Timer to measure importing audio
 	MSTimer timer;
 	timer.Start();
 
+	//Convert Wavs to Ogg's
+	std::string extension = FileDialog::GetFileExtension(filePath);
+
+	if (extension == WAV_AUDIO_EXTENSION) {
+		std::string fileIn(filePath);
+		std::string fileOut = fileIn.replace(fileIn.end() - 4, fileIn.end(), "_Converted.ogg");
```
Comment (lowysole) [inline Project/Source/FileSystem/AudioImporter.cpp:30]: Better to change only the extension, w/o _Converted

**Review by lowysole — CHANGES_REQUESTED** [review]

File: Project/Source/Components/ComponentAudioSource.cpp [hunk 9]
```diff
@@ -156,6 +159,18 @@ void ComponentAudioSource::OnEditorUpdate() {
 
 	ImGui::Separator();
 
+	if (ImGui::DragFloat("Roll Off Factor", &rollOffFact, 1.f, 0.0f, inf)) {
+		alSourcef(sourceId, AL_ROLLOFF_FACTOR, rollOffFact);
```
Comment (lowysole) [inline Project/Source/Components/ComponentAudioSource.cpp:163]: You're using rollOffFact for the three cases as variable

File: Project/Source/Components/ComponentAudioSource.cpp [hunk 10]
```diff
@@ -201,6 +216,9 @@ void ComponentAudioSource::UpdateSourceParameters() {
 	} else {
 		alSourcef(sourceId, AL_GAIN, gain);
 	}
+	alSourcef(sourceId, AL_ROLLOFF_FACTOR, rollOffFact);
```
Comment (lowysole) [inline Project/Source/Components/ComponentAudioSource.cpp:219]: Same, using rollOffFact only

**Review by lowysole — CHANGES_REQUESTED** [review]

**Review by lowysole — APPROVED** [review]

**Pull request merged**